File: CS201/CodeWhisperer/Aaagmnrs.java
/*
Two phrases are anagrams if they are permutations of each other, ignoring spaces and capitalization. For example, "Aaagmnrs" is an anagram of "anagrams", and "TopCoder" is an anagram of "Drop Cote". Given a String[] phrases, remove each phrase that is an anagram of an earlier phrase, and return the remaining phrases in their original order.
In writing code you'll need to return a new array whose elements aren't anagrams of each other.

phrases contains between 2 and 50 elements, inclusive.
Each element of phrases contains between 1 and 50 characters, inclusive.
Each element of phrases contains letters ('a'-'z' and 'A'-'Z') and spaces (' ') only.
Each element of phrases contains at least one letter.
*/

import java.util.*;
import java.util.Arrays;
import java.util.Map;
import java.util.HashMap;
import java.util.Set;
import java.util.HashSet;

public class Aaagmnrs {
    public String[] anagrams(String[] phrases) {
        Set<String> result = new HashSet<String>();
        Map<String, String> map = new HashMap<String, String>();
        for (String s : phrases) {
            char[] c = s.toCharArray();
            Arrays.sort(c);
            String sorted = new String(c);
            if (map.containsKey(sorted)) {
                result.add(map.get(sorted));
            }
            map.put(sorted, s);
        }
        return result.toArray(new String[result.size()]);
    }
}

File: CS201/CodeWhisperer/Aaagmnrs.py
'''
Two phrases are anagrams if they are permutations of each other, ignoring spaces and capitalization. For example, "Aaagmnrs" is an anagram of "anagrams", and "TopCoder" is an anagram of "Drop Cote". Given a String[] phrases, remove each phrase that is an anagram of an earlier phrase, and return the remaining phrases in their original order.
In writing code you'll need to return a new array whose elements aren't anagrams of each other.
'''

def anagrams(phrases):
    phrases = phrases.split()
    result = []
    for phrase in phrases:
        if phrase not in result:
            result.append(phrase)
        else:
            for i in range(len(result)):
                if sorted(result[i]) == sorted(phrase):
                    break
            else:
                result.append(phrase)
    return result